### PR TITLE
feat(ELB): elb policy support priority and redirect to url and fixed response

### PR DESCRIPTION
**Pull request by houpeng80**
**What this PR does / why we need it**:
  elb policy support priority and redirect to url and fixed response
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
  elb policy support priority and redirect to url and fixed response
```

## PR Checklist

* [ ] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/elb' TESTARGS='-run TestAccElbV3L7Policy_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/elb -v -run TestAccElbV3L7Policy_ -timeout 360m -parallel 4
=== RUN   TestAccElbV3L7Policy_basic
=== PAUSE TestAccElbV3L7Policy_basic
=== RUN   TestAccElbV3L7Policy_listener
=== PAUSE TestAccElbV3L7Policy_listener
=== RUN   TestAccElbV3L7Policy_url
=== PAUSE TestAccElbV3L7Policy_url
=== RUN   TestAccElbV3L7Policy_fixed_response
=== PAUSE TestAccElbV3L7Policy_fixed_response
=== CONT  TestAccElbV3L7Policy_basic
=== CONT  TestAccElbV3L7Policy_url
=== CONT  TestAccElbV3L7Policy_listener
=== CONT  TestAccElbV3L7Policy_fixed_response
--- PASS: TestAccElbV3L7Policy_url (115.21s)
--- PASS: TestAccElbV3L7Policy_fixed_response (117.48s)
--- PASS: TestAccElbV3L7Policy_basic (119.28s)
--- PASS: TestAccElbV3L7Policy_listener (133.99s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/elb       134.066s
```

### CODE REVIEW

File: docs/resources/elb_l7policy.md [hunk 1]
```diff
@@ -30,30 +101,111 @@ The following arguments are supported:
...
+  the port of the request will be used.
+
+* `path` - (Optional, String) The path that requests are redirected to. The value can contain only letters, digits,
+  and special characters _~';@^- %#&$.*+?,=!:|/()[]{} and must start with a slash (/).
+  Defaults to **${path}**, indicating that the path of the request will be used.
+
+* `query` - (Optional, String) The query string set in the URL for redirection. The value is case-sensitive and can
+  contain only letters, digits, and special characters !$&'()*+,-./:;=?@^_`. Defaults to **${query}**, indicating that
+  the query string of the request will be used.
+  For example, in the [URL](https://www.xxx.com:8080/elb?type=loadbalancer), **${query}** indicates
```
Comment (Jason-Zhang9309) [inline docs/resources/elb_l7policy.md]: this URL is an example, not a real link, so better to find another for this

File: docs/resources/elb_l7policy.md [hunk 2]
```diff
@@ -30,30 +101,111 @@ The following arguments are supported:
...
+  **type=loadbalancer**. If this parameter is set to **${query}&name=my_name**, the URL will be redirected to
+  [URL](https://www.xxx.com:8080/elb?type=loadbalancer&name=my_name).
```
Comment (Jason-Zhang9309) [inline docs/resources/elb_l7policy.md]: diito

File: docs/resources/elb_l7policy.md [hunk 3]
```diff
@@ -30,30 +101,111 @@ The following arguments are supported:
...
+  characters !$&'()+,-./:;=?@^_`. Defaults to **${query}**, indicates inherit original value.
+
+<a name="redirect_url_config_object"></a>
+The `redirect_url_config` block supports:
+
+* `status_code` - (Required, String) The status code returned after the requests are redirected.
+  Value options: **301**, **302**, **303**, **307**, **308**.
+
+* `protocol` - (Optional, String) The protocol for redirection. Value options: **HTTP**, **HTTPS**, **${protocol}**.
+  Defaults to **${protocol}**.
```
Comment (Jason-Zhang9309) [inline docs/resources/elb_l7policy.md]:
```suggestion
  Defaults to **${protocol}**, indicating that the protocol of the request will be used.
```

File: huaweicloud/services/elb/resource_huaweicloud_elb_l7policy.go [hunk 4]
```diff
@@ -114,6 +246,67 @@ func resourceL7PolicyV3Create(ctx context.Context, d *schema.ResourceData, meta
...
+	log.Printf("[DEBUG] redirectPoolsExtendConfig: %+v", redirectPoolsExtendConfig)
+	return redirectPoolsExtendConfig
+}
+
+func buildRewriteUrlConfig(data interface{}) *l7policies.RewriteUrlConfig {
+	var rewriteUrlConfig *l7policies.RewriteUrlConfig
+	rewriteUrlConfigRaw := data.([]interface{})
+	log.Printf("[DEBUG] rewriteUrlConfigRaw: %+v", rewriteUrlConfigRaw)
+	if len(rewriteUrlConfigRaw) == 1 {
+		if v, ok := rewriteUrlConfigRaw[0].(map[string]interface{}); ok {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/elb/resource_huaweicloud_elb_l7policy.go:268]: this may raise panic

File: docs/resources/elb_l7policy.md (hunk 1 above)
Comment (houpeng80) [reply docs/resources/elb_l7policy.md]: fixed

File: docs/resources/elb_l7policy.md (hunk 2 above)
Comment (houpeng80) [reply docs/resources/elb_l7policy.md]: fixed

File: docs/resources/elb_l7policy.md (hunk 3 above)
Comment (houpeng80) [reply docs/resources/elb_l7policy.md]: done

**Comment by houpeng80** [comment]
make testacc TEST='./huaweicloud/services/acceptance/elb' TESTARGS='-run TestAccElbV3L7Policy_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/elb -v -run TestAccElbV3L7Policy_ -timeout 360m -parallel 4
=== RUN   TestAccElbV3L7Policy_basic
=== PAUSE TestAccElbV3L7Policy_basic
=== RUN   TestAccElbV3L7Policy_listener
=== PAUSE TestAccElbV3L7Policy_listener
=== RUN   TestAccElbV3L7Policy_url
=== PAUSE TestAccElbV3L7Policy_url
=== RUN   TestAccElbV3L7Policy_fixed_response
=== PAUSE TestAccElbV3L7Policy_fixed_response
=== CONT  TestAccElbV3L7Policy_basic
=== CONT  TestAccElbV3L7Policy_url
=== CONT  TestAccElbV3L7Policy_fixed_response
=== CONT  TestAccElbV3L7Policy_listener
--- PASS: TestAccElbV3L7Policy_url (92.95s)
--- PASS: TestAccElbV3L7Policy_fixed_response (93.79s)
--- PASS: TestAccElbV3L7Policy_basic (98.51s)
--- PASS: TestAccElbV3L7Policy_listener (101.59s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/elb       101.660s

File: huaweicloud/services/elb/resource_huaweicloud_elb_l7policy.go [hunk 5]
```diff
@@ -114,6 +246,67 @@ func resourceL7PolicyV3Create(ctx context.Context, d *schema.ResourceData, meta
...
+			rewriteUrlConfig = &l7policies.RewriteUrlConfig{}
+			rewriteUrlConfig.Host = v["host"].(string)
+			rewriteUrlConfig.Path = v["path"].(string)
+			rewriteUrlConfig.Query = v["query"].(string)
```
Comment (ShiChangkuo) [inline huaweicloud/services/elb/resource_huaweicloud_elb_l7policy.go]:
```suggestion
			rewriteUrlConfig = &l7policies.RewriteUrlConfig{
				Host: v["host"].(string),
				Path: v["path"].(string),
				Query: v["query"].(string),
			}
```
please update them in other places

File: docs/resources/elb_l7policy.md [hunk 6]
```diff
@@ -30,30 +101,111 @@ The following arguments are supported:
 
 * `name` - (Optional, String) Human-readable name for the L7 Policy. Does not have to be unique.
 
+* `priority` - (Optional, Int) The forwarding policy priority. A smaller value indicates a higher priority. The value
+  must be unique for forwarding policies of the same listener. This parameter will take effect only when
+  `enhance_l7policy_enable` of the listener is set to **true**.
+  + If `action` is set to **REDIRECT_TO_LISTENER**, the value can only be 0.
```
Comment (ShiChangkuo) [inline docs/resources/elb_l7policy.md]: The list is unnecessary

File: huaweicloud/services/elb/resource_huaweicloud_elb_l7policy.go [hunk 7]
```diff
@@ -114,6 +246,67 @@ func resourceL7PolicyV3Create(ctx context.Context, d *schema.ResourceData, meta
 	return resourceL7PolicyV3Read(ctx, d, meta)
 }
 
+func buildRedirectPoolsExtendConfig(d *schema.ResourceData) *l7policies.RedirectPoolsExtendConfig {
+	var redirectPoolsExtendConfig *l7policies.RedirectPoolsExtendConfig
+	redirectPoolsExtendConfigRaw := d.Get("redirect_pools_extend_config").([]interface{})
+	log.Printf("[DEBUG] redirectPoolsExtendConfigRaw: %+v", redirectPoolsExtendConfigRaw)
+	if len(redirectPoolsExtendConfigRaw) == 1 {
+		redirectPoolsExtendConfig = &l7policies.RedirectPoolsExtendConfig{}
+		rewriteUrlConfigRaw := redirectPoolsExtendConfigRaw[0].(map[string]interface{})
```
Comment (ShiChangkuo) [inline huaweicloud/services/elb/resource_huaweicloud_elb_l7policy.go]: it's better to use comma-ok format

Comment (houpeng80) [reply huaweicloud/services/elb/resource_huaweicloud_elb_l7policy.go]: done

File: huaweicloud/services/elb/resource_huaweicloud_elb_l7policy.go (hunk 5 above)
Comment (houpeng80) [reply huaweicloud/services/elb/resource_huaweicloud_elb_l7policy.go]: done

File: docs/resources/elb_l7policy.md (hunk 6 above)
Comment (houpeng80) [reply docs/resources/elb_l7policy.md]: fixed

**Pull request merged**